README.md
# K8s-Watchdog

> [!WARNING]  
**Attention Users:** This project is in active development, and certain tools or features might still be under construction. We kindly urge you to exercise caution while utilizing the tools within this environment. While every effort is being made to ensure the stability and reliability of the project, there could be unexpected behaviors or limited functionalities in some areas.
We highly recommend thoroughly testing the project in non-production or sandbox environments before implementing it in critical or production systems. Your feedback is invaluable to us; if you encounter any issues or have suggestions for improvement, please feel free to [report them](https://github.com/seriohub/velero-watchdog/issues). Your input helps us enhance the project's performance and user experience.
Thank you for your understanding and cooperation.

## Description

This project is designed for monitoring the health of Kubernetes environment and alert when something is not working.
The scope is to check if the status of the elements are  in ready, running, bound state or if the amount of pods requested is always satisfied. 
Property such as CPU or memory usage are not checked to trigger an alarm and/or warning.

The user chooses the synchronization cycles and the elements to monitor.

## Features

### 1. Monitoring nodes status

Monitor both masters and works nodes.

### 2. Monitoring Pods and Workload

Monitor the status of :

- Pod
- Deployments
- StatefulSets
- DaemonSet
- ReplicaSet

### 3. Monitor Persistent Volumes and  Persistent Volume Claims

Monitor the status of :

- Persistent Volume
- Persistent Volume Claim  

### 4. Channels notifications

Receive the alerts and the solved messages via notifications channels, allowing immediate action.

Available plugin:
- Telegram
- email


## Requirements

- Python 3.x
- kubectl cli (if [Run in kubernetes](#run-in-kubernetes))
- Telegram API credentials (if telegram notification is enabled)
- SMTP and user account (if email notification is enabled)

## Configuration

| FIELD                       | TYPE   | DEFAULT | DESCRIPTION                                                                                                                                              |
|-----------------------------|--------|---------|----------------------------------------------------------------------------------------------------------------------------------------------------------|
| `DEBUG`                     | Bool   | False   | View debugging information.                                                                                                                              |
| `LOG_SAVE`                  | Bool   | False   | Save log to files                                                                                                                                        |
| `PROCESS_LOAD_KUBE_CONFIG`* | Bool   | True    | Set False if it runs on k8s.                                                                                                                             |
| `PROCESS_KUBE_CONFIG`       | String |         | Path to the kube config file. This is mandatory when the script runs outside the Kubernetes cluster, either in a docker container or as a native script. |
| `PROCESS_CLUSTER_NAME` * ** | String |         | Force the cluster name and it appears in the telegram message                                                                                            |
| `PROCESS_CYCLE_SEC`         | Int    | 120     | Cycle time (seconds)                                                                                                                                     |
| `TELEGRAM_ENABLE`    *      | Bool   | True    | Enable telegram notification                                                                                                                             |
| `TELEGRAM_API_TOKEN` *      | String |         | Token for access to Telegram bot via Http API                                                                                                            |
| `TELEGRAM_CHAT_ID`   *      | String |         | Telegram chat id where send the notifications                                                                                                            |
| `EMAIL_ENABLE`       *      | Bool   | False   | Enable email notification                                                                                                                                |
| `EMAIL_SMTP_SERVER`  *      | String |         | SMTP server                                                                                                                                              |
| `EMAIL_SMTP_PORT`    *      | int    | 587     | SMTP port                                                                                                                                                |
| `EMAIL_ACCOUNT`      *      | String |         | user name account                                                                                                                                        |
| `EMAIL_PASSWORD`     *      | String |         | password account                                                                                                                                         |
| `EMAIL_RECIPIENTS`   *      | Bool   |         | Email recipients                                                                                                                                         |
| `K8S_NODES`                 | Bool   | True    | Enable Nodes watcher                                                                                                                                     |
| `K8S_PODS`                  | Bool   | True    | Enable Pods watcher                                                                                                                                      |
| `K8S_DEPLOYMENT`            | Bool   | True    | Enable Deployment watcher                                                                                                                                |
| `K8S_STATEFUL_SETS`         | Bool   | True    | Enable StatefulSets watcher                                                                                                                              |
| `K8S_REPLICA_SETS`          | Bool   | True    | Enable ReplicaSets watcher                                                                                                                               |
| `K8S_DAEMON_SETS`           | Bool   | True    | Enable DaemonSet watcher                                                                                                                                 |
| `K8S_PVC`                   | Bool   | True    | Enable Persistent Volume Claim watcher                                                                                                                   |
| `K8S_PV`                    | Bool   | True    | Enable Persistent Volume watcher                                                                                                                         |
| `K8S_DEPLOYMENT_P0`         | Bool   | False   | Load the Deployment items where the pods count is 0                                                                                                      |
| `K8S_STATEFUL_SETS_P0`      | Bool   | False   | Load the StatefulSets items where the pod count is 0                                                                                                     |
| `K8S_REPLICA_SETS_P0`       | Bool   | False   | Load the ReplicaSets items where the pod count is 0                                                                                                      |
| `K8S_DAEMON_SETS_P0`        | Bool   | False   | Load DaemonSet items where the pods count is scaled to 0 or is equal to 0                                                                                |

*Mandatory parameters<br>
** Mandatory if it is deployed on cluster

<br>If you set "TELEGRAM_ENABLE"= False, the application prints out the message only on the stdout

## Installation

Clone the repository:

  ``` bash
    git clone https://github.com/seriohub/k8s-watchdog.git
    cd k8s-watchdog
  ```

### Run native

1. Navigate to the [src](src) folder

2. Dependencies installation:

    ``` bash
    pip install -r requirements.txt
    ```

3. Configuration

    Create and edit .env file under src folder, you can start from [.env.template](src/.env.template) under [src](src) folder
    Setup mandatory parameters in the src/.env file if runs it in the native mode

4. Usage

    Run the main script:

    ``` bash
    python3 main.py
    ```

### Run in Docker

1. Configuration
   1. Navigate to the [docker](docker) folder
   2. Setup mandatory parameters in the docker-compose.yaml file ([docker-compose.yaml](docker/docker-compose.yaml))

      **Note:** Instead of editing the docker-compose.yaml file, you can create and edit .env file (you can start from [.env.template](docker/.env.template)) file under [docker](docker) folder and use the docker-compose.yaml default  values.
2. Docker image :
   ##### Build docker image from scratch
      1. Navigate to the root folder
      2. Build the docker image
      ``` bash
      docker build --target k8s-watchdog -t k8s-watchdog:1.0.0 -t k8s-watchdog:latest -f ./docker/Dockerfile .
      ```
   ##### Use image published on DockerHub
      1. Pull the image 
      ``` bash
      docker pull dserio83/k8s-watchdog
      ```
3. Create docker volumes:
   1. Create a volume for store logs
      ``` bash
      docker volume create k8s_watchdog_vol
      ```
   2. Create a volume for config data.
      ``` bash
      docker volume create k8s_watchdog_config
      ```
4. Copy the kube config file into the volume created. It contains the credentials for the k8s API.
Generally it is in /home/< user >/.kube/config folder.

   ``` bash
   docker container create --name copy-file -v k8s_watchdog_config:/config hello-world
   docker cp <path config file yaml> copy-file:/config/config.yaml
   docker rm copy-file
   ```
5. Create the stack and run it

    ``` bash
     docker-compose  -f ./docker/docker-compose.yaml -p k8s-watchdog-stack  up -d
    ```

### Run in Kubernetes

1. Configuration
   1. Navigate to the [k8s](k8s) folder
   2. Create and edit .env file under k8s folder, you can start from [.env.template](k8s/.env.template)
  
   3. Export .env

      ``` bash
      export $(cat .env | xargs)
      ```

      Checks that the variables have been exported:

      ``` bash
      printenv | grep K8SW_* 
      ```
 
2. Setup docker image:

   1. Navigate to the root folder
   2. Build image

        ``` bash
        docker build --target k8s-watchdog -t ${K8SW_DOCKER_REGISTRY}/${K8SW_DOCKER_IMAGE}:1.0.0 -t ${K8SW_DOCKER_REGISTRY}/${K8SW_DOCKER_IMAGE}:latest -f ./docker/Dockerfile .
        ```

   3. Push image

        ``` bash
        docker push ${K8SW_DOCKER_REGISTRY}/${K8SW_DOCKER_IMAGE} --all-tags
        ```

      >[!INFO]  
      Alternative you can use skip the *Build image* and *Push image* steps and use an deployed image published on DockerHub.<br>
      Edit the .env file:
      **K8SW_DOCKER_REGISTRY=dserio83** <br>
      More info: https://hub.docker.com/r/dserio83/k8s-watchdog

3. Kubernetes create objects

   1. Navigate to the [k8s](k8s) folder

   2. Create namespace:

        ``` bash
        cat 10_create_ns.yaml | envsubst | kubectl apply -f -
        ```

   3. Create the PVC:

       ``` bash
       cat 20_pvc.yaml | envsubst | kubectl apply -f -
       ```

   4. Create the ConfigMap:

       ``` bash
       cat 30_cm.yaml | envsubst | kubectl apply -f -
       ```
  
   5. Create the RBAC:

       ``` bash
        cat 40_rbac.yaml | envsubst | kubectl apply -f -
       ```
  
   6. Create the Deployment:

       ``` bash
        cat 50_deployment.yaml | envsubst | kubectl apply -f -
       ```

## Test Environment

The project is developed, tested and put into production on several clusters with the following configuration

1. Kubernetes v1.28.2

## How to Contribute

1. Fork the project
2. Create your feature branch

    ``` bash
    git checkout -b feature/new-feature
    ```

3. Commit your changes

    ``` bash
   git commit -m 'Add new feature'
   ```

4. Push to the branch

    ``` bash
   git push origin feature/new-feature
   ```

5. Create a new pull request

## License

This project is licensed under the [MIT License](LICENSE).

---

Feel free to modify this template according to your project's specific requirements.

In case you need more functionality, create a PR. If you find a bug, open a ticket.


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file.

## [0.2.1] - 2023-11-06
**Fixed bugs:**
- The same content was written several times in the alive message 


## [0.2.0] - 2023-11-03
**Implemented enhancements:**
- Added mechanism to handle more channels of notifications
- Added email channel notification
- Added mechanism for optimizing the output message


## [0.1.5] - 2023-10-29
**Implemented enhancements:**
- In the alive message added the name of the cluster and the time without warning/alarm
- In pods filters, added the status "ContainerCreating" to the OK


## [0.1.4] - 2023-10-28
**Implemented enhancements:**
- The response from http request (telegram api call)  is limited to 10 characters
- In pod message details add the container status

**Fixed bugs:**
- The pod in "Succeeded" state is not considered a critical state  

## [0.1.3] - 2023-10-25

**Implemented enhancements:**
- add a new key in .env file to define the cluster name. When the script runs in the k8s, get cluster name does not work 
- optimized source code in file main.py 
- add version.py to store the version of script. It is aligned to the CHANGELOG.md latest version

## [0.1.2] - 2023-10-23
 
**Implemented enhancements:**
- add Jenkins pipeline for building the docker image and push it to the hub registry 

## [0.1.1] - 2023-10-21
 
**Implemented enhancements:**
- add cluster name in the message details

**Fixed bugs:**
- during the loading of .env keys, the system printed wrong default value

## [0.1.0] - 2023-10-20
 
### First Commit

v0.1.0: [https://github.com/seriohub/k8s-watchdog/releases/tag/v0.1.0](https://github.com/seriohub/k8s-watchdog/releases/tag/v0.1.0)

src/utils/config.py
from dotenv import load_dotenv
import os
from utils.handle_error import handle_exceptions_static_method, handle_exceptions_method


# class syntax
class ConfigProgram:
    def __init__(self, path_env=None, debug_on=True):
        self.debug_on = debug_on
        res = load_dotenv(dotenv_path=path_env)
        print(f"INFO    [Env] Load env={res}")

    @handle_exceptions_static_method
    def load_key(self, key, default, print_out: bool = True, mask_value: bool = False):
        value = os.getenv(key)
        if value is None or \
                len(value) == 0:
            value = default

        if print_out:
            if mask_value and len(value) > 2:
                index = int(len(value) / 2)
                partial = '*' * index
                if self.debug_on:
                    print(f"INFO    [Env] load_key.key={key} value={value[:index]}{partial}")
            else:
                if self.debug_on:
                    print(f"INFO    [Env] load_key.key={key} value={value}")

        return value

    @handle_exceptions_method
    def logger_key(self):
        return self.load_key('LOG_KEY', 'k8s-wdt')

    @handle_exceptions_method
    def logger_msg_format(self):
        default = '%(asctime)s :: [%(levelname)s] :: %(message)s'
        return self.load_key('LOG_FORMAT', default)

    @handle_exceptions_method
    def logger_save_to_file(self):
        res = self.load_key('LOG_SAVE', 'False')
        return True if res.upper() == 'TRUE' else False

    @handle_exceptions_method
    def logger_folder(self):
        return self.load_key('LOG_DEST_FOLDER',
                             './logs')

    @handle_exceptions_method
    def logger_filename(self):
        return self.load_key('LOG_FILENAME',
                             'k8s.log')

    @handle_exceptions_method
    def logger_max_filesize(self):
        return int(self.load_key('LOG_MAX_FILE_SIZE',
                                 4000000))

    @handle_exceptions_method
    def logger_his_backups_files(self):
        return int(self.load_key('LOG_FILES_BACKUP',
                                 '5'))

    @handle_exceptions_method
    def logger_level(self):
        return int(self.load_key('LOG_LEVEL',
                                 '20'))

    @handle_exceptions_method
    def process_run_sec(self):
        res = self.load_key('PROCESS_CYCLE_SEC',
                            '120')

        if len(res) == 0:
            res = '120'
        return int(res)

    @handle_exceptions_method
    def internal_debug_enable(self):
        res = self.load_key('DEBUG', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def telegram_token(self):
        return self.load_key('TELEGRAM_TOKEN', '', mask_value=True)

    @handle_exceptions_method
    def telegram_chat_id(self):
        return self.load_key('TELEGRAM_CHAT_ID', '', mask_value=True)

    @handle_exceptions_method
    def telegram_enable(self):
        res = self.load_key('TELEGRAM_ENABLE', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def telegram_max_msg_len(self):
        res = self.load_key('TELEGRAM_MAX_MSG_LEN',
                            '3000')

        if len(res) == 0:
            res = '2000'
        return int(res)

    @handle_exceptions_method
    def telegram_rate_limit_minute(self):
        res = self.load_key('TELEGRAM_MAX_MSG_MINUTE',
                            '20')

        if len(res) == 0:
            res = '20'
        return int(res)

    @handle_exceptions_method
    def notification_alive_message_hours(self):
        res = self.load_key('NOTIFICATION_ALIVE_MSG_HOURS',
                            '24')
        n_hours = int(res)
        if n_hours < 0:
            n_hours = 0
        elif n_hours > 100:
            n_hours = 100

        return n_hours

    @handle_exceptions_method
    def email_enable(self):
        res = self.load_key('EMAIL_ENABLE', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def email_smtp_server(self):
        return self.load_key('EMAIL_SMTP_SERVER', '')

    @handle_exceptions_method
    def email_smtp_port(self):
        res = self.load_key('EMAIL_SMTP_PORT',
                            '587')
        n_port = int(res)
        return n_port

    @handle_exceptions_method
    def email_recipient(self):
        return self.load_key('EMAIL_RECIPIENTS', '')

    @handle_exceptions_method
    def email_sender_password(self):
        return self.load_key('EMAIL_PASSWORD', '', mask_value=True)

    @handle_exceptions_method
    def email_sender(self):
        return self.load_key('EMAIL_ACCOUNT', '')

    @handle_exceptions_method
    def k8s_load_kube_config_method(self):
        res = self.load_key('PROCESS_LOAD_KUBE_CONFIG', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_config_file(self):
        return self.load_key('PROCESS_KUBE_CONFIG', None)

    @handle_exceptions_method
    def k8s_force_cluster_identification(self):
        return self.load_key('PROCESS_CLUSTER_NAME', None)

    @handle_exceptions_method
    def k8s_nodes_enable(self):
        res = self.load_key('K8S_NODE', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_pods_enable(self):
        res = self.load_key('K8S_PODS', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_deployment_enable(self):
        res = self.load_key('K8S_DEPLOYMENT', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_deployment_pods0(self):
        res = self.load_key('K8S_DEPLOYMENT_P0', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_stateful_sets_enable(self):
        res = self.load_key('K8S_STATEFUL_SETS', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_stateful_sets_pods0(self):
        res = self.load_key('K8S_STATEFUL_SETS_P0', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_replica_sets_enable(self):
        res = self.load_key('K8S_REPLICA_SETS', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_replica_sets_pods0(self):
        res = self.load_key('K8S_REPLICA_SETS_P0', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_daemons_sets_enable(self):
        res = self.load_key('K8S_DAEMON_SETS', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_daemons_sets_pods0(self):
        res = self.load_key('K8S_DAEMON_SETS_P0', 'False')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_pvc_enable(self):
        res = self.load_key('K8S_PVC', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False

    @handle_exceptions_method
    def k8s_pv_enable(self):
        res = self.load_key('K8S_PV', 'True')
        return True if res.lower() == "true" or res.lower() == "1" else False


class ConfigK8sProcess:
    def __init__(self, cl_config: ConfigProgram = None):
        self.CLUSTER_Name_forced = None
        self.CLUSTER_Name_key = 'cluster'

        self.POD_enable = True
        self.POD_key = 'pods'

        self.SS_enable = True
        self.SS_pods0 = False
        self.SS_key = 'stateful_sets'

        self.RS_enable = True
        self.RS_pods0 = False
        self.RS_key = 'replicaset_sets'

        self.DS_enable = True
        self.DS_pods0 = True
        self.DS_key = 'daemon_sets'

        self.PVC_enable = True
        self.PVC_key = 'pvc'

        self.PV_enable = True
        self.PV_key = 'pv'

        self.DPL_enable = True
        self.DPL_pods0 = False
        self.DPL_key = 'deployment'

        self.NODE_enable = True
        self.NODE_key = 'nodelist'

        # LS 2023.11.03 key for sending a unique message
        self.disp_MSG_key_unique = True  # Fixed True
        self.disp_MSG_key_start = 'msg_key_start'
        self.disp_MSG_key_end = 'msg_key_end'

        if cl_config is not None:
            self.__init_configuration_app__(cl_config)

    def __print_configuration__(self):
        """
        Print setup class
        """
        if self.CLUSTER_Name_forced is not None:
            print(f"INFO    [Process setup] k8s cluster name= {self.CLUSTER_Name_forced}")

        print(f"INFO    [Process setup] k8s check node={self.NODE_enable}")
        print(f"INFO    [Process setup] k8s check daemon sets={self.DS_enable}- pods0={self.DS_pods0}")
        print(f"INFO    [Process setup] k8s check pods ={self.POD_enable}")
        print(f"INFO    [Process setup] k8s check deployment={self.DPL_enable}- pods0={self.DPL_pods0}")
        print(f"INFO    [Process setup] k8s check stateful sets={self.SS_enable}- pods0={self.SS_pods0}")
        print(f"INFO    [Process setup] k8s check replicaset={self.RS_enable}- pods0={self.RS_pods0}")
        print(f"INFO    [Process setup] k8s check pvc={self.PVC_enable}")
        print(f"INFO    [Process setup] k8s check pv={self.PVC_enable}")

        print(f"INFO    [Process setup] k8s send summary message={self.disp_MSG_key_unique}")

    def __init_configuration_app__(self, cl_config: ConfigProgram):
        """
        Init configuration class reading .env file
        """
        self.PVC_enable = cl_config.k8s_pvc_enable()
        self.PV_enable = cl_config.k8s_pv_enable()
        self.POD_enable = cl_config.k8s_pods_enable()
        self.NODE_enable = cl_config.k8s_nodes_enable()
        self.DS_enable = cl_config.k8s_daemons_sets_enable()
        self.DPL_enable = cl_config.k8s_deployment_enable()
        self.SS_enable = cl_config.k8s_stateful_sets_enable()
        self.RS_enable = cl_config.k8s_replica_sets_enable()
        self.DS_pods0 = cl_config.k8s_daemons_sets_pods0()
        self.DPL_pods0 = cl_config.k8s_deployment_pods0()
        self.SS_pods0 = cl_config.k8s_stateful_sets_pods0()
        self.RS_pods0 = cl_config.k8s_replica_sets_pods0()
        self.CLUSTER_Name_forced = cl_config.k8s_force_cluster_identification()

        self.__print_configuration__()


class ConfigDispatcher:
    def __init__(self, cl_config: ConfigProgram = None):
        self.max_msg_len = 50000
        self.alive_message = 24

        self.telegram_enable = False
        self.telegram_chat_id = '0'
        self.telegram_token = ''
        self.telegram_max_msg_len = 2000
        self.telegram_rate_limit = 20

        self.email_enable = True
        self.email_smtp_server = ''
        self.email_smtp_port = 587
        self.email_sender = ''
        self.email_sender_password = '***'
        self.email_recipient = ''

        if cl_config is not None:
            self.__init_configuration_app__(cl_config)

    def __mask_data__(self, data):
        if len(data) > 2:
            index = int(len(data) / 2)
            partial = '*' * index
            return f"{self.telegram_token[:index]}{partial}"
        else:
            return data

    def __print_configuration__(self):
        """
        Print setup class
        """

        print(f"INFO    [Dispatcher setup] telegram={self.telegram_enable}")
        if self.telegram_enable:
            print(f"INFO    [Dispatcher setup] telegram-chat id={self.telegram_chat_id}")
            # if len(self.telegram_token) > 2:
            #     index = int(len(self.telegram_token) / 2)
            #     partial = '*' * index
            #     print(f"INFO    [Dispatcher setup] telegram-token={self.telegram_token[:index]}{partial}")
            # else:
            #     print(f"INFO    [Dispatcher setup] telegram-token={self.telegram_token}")
            print(f"INFO    [Dispatcher setup] telegram-token={self.__mask_data__(self.telegram_token)}")

            print(f"INFO    [Dispatcher setup] telegram-max message length={self.telegram_max_msg_len}")
            print(f"INFO    [Dispatcher setup] telegram-rate limit minute={self.telegram_rate_limit}")
            print(f"INFO    [Dispatcher setup] Notification-alive message every={self.alive_message} hour")

        print(f"INFO    [Dispatcher setup] email={self.email_enable}")
        if self.email_enable:
            print(f"INFO    [Dispatcher setup] email-smtp server={self.email_smtp_server}")
            print(f"INFO    [Dispatcher setup] email-port={self.email_smtp_port}")
            print(f"INFO    [Dispatcher setup] email-sender={self.email_sender}")
            # print(f"INFO    [Dispatcher setup] email-password={self.__mask_data__(self.email_sender_password)}")
            print(f"INFO    [Dispatcher setup] email-password={self.email_sender_password}")
            print(f"INFO    [Dispatcher setup] email-recipient={self.email_recipient}")

    def __init_configuration_app__(self, cl_config: ConfigProgram):
        """
        Init configuration class reading .env file
        """
        # telegram section
        self.telegram_enable = cl_config.telegram_enable()
        self.telegram_chat_id = cl_config.telegram_chat_id()
        self.telegram_token = cl_config.telegram_token()
        self.telegram_max_msg_len = cl_config.telegram_max_msg_len()
        self.telegram_rate_limit = cl_config.telegram_rate_limit_minute()

        self.email_enable = cl_config.email_enable()
        self.email_sender = cl_config.email_sender()
        self.email_sender_password = cl_config.email_sender_password()

        self.email_smtp_port = cl_config.email_smtp_port()
        self.email_smtp_server = cl_config.email_smtp_server()
        self.email_recipient = cl_config.email_recipient()

        self.alive_message = cl_config.notification_alive_message_hours()

        # email section
        self.__print_configuration__()
